### ISSUE TITLE
HTTP/2 support

### ISSUE THREAD

**Issue by TheLonelyGhost** [manastech/crystal#2125]
This is a feature request, tag it as you will. Since the specs have been clearly defined and do not appear to waver, I feel HTTP/2 should be supported in addition to HTTP/1.x.

Resources:

- [HTTP2 Working Group](https://http2.github.io/) - includes specs and running list of implementations in other languages
- [Ruby HTTP2 gem](https://github.com/igrigorik/http-2) - Maintained by Ilya Grigorik for both client and server

**Comment by asterite** [comment]
Thanks for adding this, we'll eventually need to support HTTP/2.

I'm not sure how it works, though. Will it integrate automagically with the current server? Because in Ilya Grigorik's implementation it doesn't look like it integrates with Rack. There's also this [long thread](https://github.com/tenderlove/the_metal/issues/5) which supports my guessing. And [this Go talk](https://talks.golang.org/2016/state-of-go.slide#9) which says that if you are using HTTPS and Go 1.6 you are using HTTP/2, so I guess it should be transparent.

Maybe @igrigorik could chime in and at least let us know if the current client/server implementation is easily upgradable to HTTP/2? :grin:

**Comment by igrigorik** [comment]
@asterite you can definitely abstract h2 behind the same API. That said, I do think you want to think carefully about the API to enable developers take advantage of h2: persistent connections, multiplexing, server push, etc. "Rack API", at least as it was defined for http/1.1, doesn't do a good job of this.

FWIW, check out the demo client/server examples in the ruby gem: 
https://github.com/igrigorik/http-2/tree/master/example

**Comment by asterite** [comment]
@igrigorik Thank you for replying, I'll take a look at the examples and see if I can port something to Crystal.

**Comment by ysbaddaden** [comment]
@asterite recently changed HTTP::Server from a Rack-like interface (get a request, return a response) to instead pass a Context object with the Request and Response objects that middlewares will change, each context being evaluated in its own coroutine.

I have a feeling that HTTP::Server could handle the HTTP/2 protocol, creating a Context object for each incoming Streams, using an intermediate IO to write to the Stream instead of the Socket —like deflate middleware currently does. The Context object could have additional methods for server push (among other HTTP/2 features), and either raise or silently skip when HTTP/2 isn't supported by the client.

Hopefully this wouldn't affect the current middleware API but enhance it.

**Comment by ysbaddaden** [comment]
An initial implementation is available: https://github.com/ysbaddaden/http2

If someone could point me to great HTTP2 compliance test suites, that would be very much appreciated.

The **client part** is untested. It should be capable to use HTTP2::Connection, but it may require some tweaks. It's integration into `HTTP::Client` may be tricky, especially since we can have many requests in parallel (yay, fibers!) and the API needs a callback to handle server-push promises (accept/refuse then receive data). We can prevent server-push by default, too. We also need to configure local HTTP2 settings (frame size, max concurrent streams, etc).

The **server part** is almost complete, except for flow control (required) and applying priorities to send frames (nice to have). It's capable to talk with nghttp (with or without TLS) and browsers (ie. Firefox and Chrome, TLS only) and it supports server-push.

Now we should start planning to integrate this into `HTTP::Server`. For example:
- [x] ~~Add `HTTP::Server.default_ssl_context` that would return an `OpenSSL::SSL::Context` that is compliant to HTTP2 requirements and always provides state-of-the-art OpenSSL configuration; `HTTP::Client` could provide the same;~~ It's now `OpenSSL::SSL::Context::Server`;
- [x] Introduce an abstraction for HTTP1 responses, to leverage the protocol differences between HTTP1 and HTTP2 so `HTTP::Server::Response::Output` doesn't have to deal with protocol details (ie. encoding HTTP headers, writing body to the socket or as a stream data frame);
- [x] Transparently support HTTP2 in `HTTP::Server`:
  - [ ] configure local HTTP2 settings (max frame size, concurrent streams, etc);
  - [x] handle the ALPN protocol for SSL connections;
  - [x] react to the `Upgrade: h2c` header for non-SSL connections (and discard `Upgrade: h2`);
  - [x] handlers will deal with `HTTP2::Stream` objects instead of a Socket, which is similar to how we wrap the socket in `HTTP::Server::DeflateHandler` (which will wrap `HTTP2::Stream` now);
- [ ] Add method(s) to `HTTP::Server::Context` to enable server-push, with a solution to actually send the response, or not, if the Client accepted or refused them —and either raise or silently discard if the connection ain't HTTP2.

**Comment by igrigorik** [comment]
@ysbaddaden take a look at https://github.com/summerwind/h2spec

**Comment by ysbaddaden** [comment]
Thank you @igrigorik! I was able to reduce a bunch of corner cases in the protocol, down to 4 failures, most related to the unimplemented flow-control.

**Comment by sdogruyol** [comment]
Hey @ysbaddaden i remember you implementing HTTP/2 for Crystal. What's the status on that 😄

**Comment by samueleaton** [comment]
I too am _VERY_ interested in the [speed improvements](https://www.troyhunt.com/i-wanna-go-fast-https-massive-speed-advantage/) that HTTP2 offers. I didn't know it had such [wide support](http://caniuse.com/#feat=http2). 

@ysbaddaden should the work you're doing for http2 be moved into a crystal branch? I wouldn't mind writing a bunch of specs for it.

**Comment by DebugReport** [comment]
Up. 
Any progress?

**Comment by spalladino** [comment]
@DebugReport not that we are aware of. @samueleaton, if you are interested in contributing, I'd advise to do so in @ysbaddaden's repo, and as it's finished and tested, we can then include it into the stdlib.

**Comment by jeromegn** [comment]
As far as I can tell @ysbaddaden's http2 shard is feature-complete and passes all h2spec tests. Awesome work.

It doesn't look like there are any plans for integrating it back into Crystal though. I was planning on using it in my own project, a shard is fine, but I think it'd be nice to have in the std lib. I don't think http2 is going away.

**Comment by luislavena** [comment]
Hello @asterite, @bcardiff, @ysbaddaden, @waj and other core-team members,

Would like to bring to your attention this issue, hoping if possible to introduce the changes required to support HTTP/2 into the stdlib.

Not necessarily suggesting introduce HTTP/2 itself to the stdlib, but perhaps tweak `HTTP::Server` and friends to avoid the differences currently required to support it, in that way, someone can load the http2 shard and use when required.

Given that we are still in pre-1.0, perhaps is a good time to do some final breaking changes to the related API in preparation for that.

HTTP/2 support will also open the door to gRPC, which can increase interoperability of Crystal for certain teams and projects. Also worth to mention HTTP/3 is around the corner :wink:

Thank you for your hard and continuous work both Crystal core and other contributors, every new release gets better and better and you can feel how much sweat and hard work has been pour into it. Thank you! ❤️ ❤️ ❤️ 

Cheers.

**Comment by RX14** [comment]
I think that `HTTP` needs a ground-up clean API redesign (the implementation is fine). It has collected a lot of issues which need to be collated into a requirements list for a `HTTP` module which is ready to be marked 1.0.

**Comment by bcardiff** [comment]
Regarding HTTP (Client/Server) we wanted a way to decouple the protocol implementation from the socket tls wrapper. That will allow simple http over unix socket and other use cases where a simple http api is needed. (And would actually remove the need for the `without_openssl` flag).

**Comment by pyrsmk** [comment]
As a side note about WebSockets working with HTTP/2, I'm dropping these articles here, for those who are interested : 

https://hpbn.co/websocket/#websocket-multiplexing-and-head-of-line-blocking
https://medium.com/@pgjones/http-2-websockets-81ae3aab36dd

And the related RFC : 

https://tools.ietf.org/html/rfc8441

**Comment by eliasjpr** [comment]
Another take at Crystal HTTP implementation https://github.com/azutoolkit/duo

**Comment by straight-shoota** [comment]
I suppose at this point HTTP/2 support isn't that useful compared to HTTP/3 (which is basically the same, just based on QUIC instead of TCP).
Recently released OpenSSL 3.2 has QUIC client support (server is expected to follow in the following release(s)), so this could serve for an easy integration into `HTTP::Client`.
That would be more approachable than an additional QUIC library (which https://github.com/iv-org/lsquic.cr is using).

**Comment by luislavena** [comment]
Recently Ilya Grigorik released a pure-Ruby implementation of HTTP/2: https://github.com/igrigorik/http-2 that decouples the transport from the HTTP/2 processing.

**Comment by dentarg** [comment]
@luislavena that gem has been around [since 2013 with the last release in 2021](https://rubygems.org/gems/http-2/versions) before this recent flurry of activity that seems to be led by AWS (https://github.com/mullermp)

There is also https://github.com/socketry/protocol-http2 by Samuel Williams

**Comment by HCLarsen** [comment]
> I suppose at this point HTTP/2 support isn't that useful compared to HTTP/3 (which is basically the same, just based on QUIC instead of TCP). 

I would say that it is just as useful, as it's possible that some clients/servers could support HTTP/2, but not HTTP/3. We should be going for all options that developers may be using.

In any case, is anyone working on a PR yet? If not, I may tackle this myself over the summer.

**Comment by straight-shoota** [comment]
You're right. I didn't realize at the time I wrote this, that there are still good reasons for HTTP/2.

I'm not aware of anyone working on this. Your help would be appreciated 👍

**Comment by ryanprior** [comment]
I had a conversation the other day about how I wish I had the background knowledge & time to take this on. A lot of tooling in the K8s/cluster orchestration space is based on HTTP2, and support for that stack in the core library would make crystal a much easier sell in the cloud ops space.

If somebody else wants to take point, I'd be happy to try and pitch in! I definitely don't have the right background to lead the effort, nor do I currently have the time to skill-up and read the relevant RFCs & existing implementations.

**Comment by HCLarsen** [comment]
Another thought about this. When making requests with `HTTP::Client`, it's entirely possible that someone may want to specify which HTTP protocol they want to use. Unless I'm missing it, there's currently no mechanism for that.

My first thought for a solution includes an enum for versions, and a constant that points to the newest version (Yes, I'm thinking forward compatibility).

**Comment by ysbaddaden** [comment]
Better late than never. I ticked some boxes in my [comment above](https://github.com/crystal-lang/crystal/issues/2125#issuecomment-211024868) from... 9 years ago 😅

I took a few minutes to fix the few issues because the last release of my [http2 shard](https://github.com/ysbaddaden/http2) was for Crystal 0.34, some more to fix a bug and find a workaround for an annoying issue in h2spec (the shard used to pass h2spec 2.1 but not later releases), then a couple days to **hack transparent support** into `HTTP::Server` ✨

That's right, adding HTTP/2 support to your HTTP servers is a mere:

```crystal
require "http2/server"
```

After adding the [shard](https://github.com/ysbaddaden/http2) obviously, and if your server or framework sticks to `HTTP::Server` of course. But then, you shouldn't have to do anything else. Oh, except configure a TLS certificate because browsers only talk HTTP/2 through HTTPS, even on localhost 😮‍💨

**Comment by ysbaddaden** [comment]
Now there are probably bugs, a couple things are missing (like max header size, max request line size), and we can probably do better/simpler in the monkey-patch (or evolve `HTTP::Server::Response` vs `Output`), WebSockets should work but I didn't test, ...

Please investigate & report if you find some, or better open a PR with a fix 🙇

**Comment by bararchy** [comment]
@ysbaddaden Does `require "http2"` will also make HTTP::Client HTTP2 compatible like it will do for the Server?

**Comment by ysbaddaden** [comment]
@bararchy That's not the case, _yet_, but eventually yes, `require "http2/client"` should monkey patch HTTP/2 support into `HTTP::Client`. I looked a bit yesterday, and that mostly boils down to patching the `#io` and `#exec_internal` methods, deal with all the possible scenarios to start a HTTP/2 connection and fallback to HTTP/1 (with potential HTTP/2 upgrade), maybe provide a mean to force a version.

Ideally, if the connection is HTTP/1 then once a HTTP client is opened, concurrent requests from multiple fibers should be sequential with a Mutex for example, while if HTTP/2 they could be multiplexed.

Then the main complexity is probably dealing with GOAWAY frames (some requests may be completing, some may have been discarded).

**Comment by eliasjpr** [comment]
@ysbaddaden happy to see this. Please let me know if you would like to read the changes in the HTTP Duo. Out of respect to you and the community and because of licensing issues I have archived and deleted the Repo. I can make the source code available if you ever need it. 

Thanks for all the work over the years!
